Add support for object expansion

swelham opened the pull request:
This PR adds basic support for object expansion as described [here](https://stripe.com/docs/api#expanding_objects) and aims to resolve #324.

I still need to add documentation/tests, however I wanted to check this is the desired way to do this before doing so.

These changes allow an `:expand` option with a list value to be passed via `opts` into any of the api functions on resources that support expansion. If an incorrect expansion is specified or one is specified on a resource that doesn't support it, the stripe api will return a 400 bad request.

Example usage

```elixir
Stripe.Charge.retrieve("ch_123", expand: [:balance_transaction, :customers])
```

or using a binary

```elixir
Stripe.Charge.retrieve("ch_123", expand: ["balance_transaction", "customers"])
```

I'm not entirely sure if having the `prefix_expansions` function in the request module is the correct place for this, any feedback on this would be great.

snewcomer commented on lib/stripe/request.ex:
```diff
@@ -144,6 +144,20 @@ defmodule Stripe.Request do
+        request
+
+      expansions ->
+        mapped_expansions = Enum.map(expansions, &"#{prefix}.#{&1}")
```
What is the need to prefix with `data`?  (sry I prob am not seeing something)

swelham replied:
The prefix is required when the stripe api is returning a list of resources as they get wrapped in a `data: {}` attribute.

snewcomer replied:
@swelham will stripe's response actually expand an expand-able item if the url looks something like this? i.e. prefixed with `data`?

`http://localhost:12123/v1/card?expand[]=data.account`

I thought it has to be like, thus the question of why the `prefix_expansion` is needed.

`http://localhost:12123/v1/card?expand[]=account`

...or for multiple

`http://localhost:12123/v1/card?expand[]=account&expand[]=customer`

I would imagine stripe's response would error...What are your thoughts?

swelham replied:
In all cases with the exception of returning a list of resources the prefix is not needed. This isn't documented anywhere in the docs as far I can tell and I worked it out by querying the api.

So for create, update and retrieve the following is correct.

`/v1/charges?expand[]=balance_transaction`

Which returns this structure.

```
{
  "id": "ch_1CeIjZC4XIOiOpjmm4mHdB0u",
  "balance_transaction": {
    ...
  },
  ...
}
```

And then for listing resources the following is correct.

`/v1/charges?expand[]=data.balance_transaction`

Which returns this structure.

```
{
  "data": [
    {
      "id": "ch_1CeIjZC4XIOiOpjmm4mHdB0u",
      "balance_transaction": {
        ...
      },
      ...
    }
  ]
}
```

In the case for multiples, yes the format of having multiple `"expand[]="` is correct with the value following the examples above. The stripe api will return a bad request if these are spelt or formatted incorrectly.

Multiples examples.

Api create, update, retrieve

`/v1/charges?expand[]=balance_transaction&expand[]=customer`

Api list.

`/v1/charges?expand[]=data.balance_transaction&expand[]=data.customer`

I hope that makes sense 😃

snewcomer replied:
Nice sleuthing!  I just curled the endpoints, and yes, you are completely correct.  💯

snewcomer replied:
I'm guessing this also means for `cancel` and `delete`?

swelham replied:
Yes for `cancel` and no for `delete`. The deletes only return the id of the resource deleted and a flag to indicate success, so there is nothing to expand here.

snewcomer commented on lib/stripe/request.ex:
```diff
@@ -144,6 +144,20 @@ defmodule Stripe.Request do
   def get_id!(_), do: raise("You must provide an ID or a struct with an ID to this operation.")
 
+  @spec prefix_expansions(t, binary) :: list
+  def prefix_expansions(%Request{opts: opts} = request, prefix) do
```
Would this be needed if the request looked like this?  I.e. it could be handled by your awesome work in `add_object_expansion`?

`Stripe.Charge.retrieve("ch_123", [expand: "balance_transaction"])`

For multiple, here is one option.  We would need to split on `,` to generate `expand[]=balance_transaction&expand[]=application_fee`.
`Stripe.Charge.retrieve("ch_123", [expand: "balance_transaction,application_fee"])`

snewcomer replied:
Note I think `expanded: ["balance_transaction"]` is ruby? whereas `opts` in our case is a keyword list (https://elixir-lang.org/getting-started/keywords-and-maps.html)...

Side note - I also noticed a doc block that could be updated.

swelham replied:
The `prefix_expansions` needs to happen before it makes it into `add_object_expansion` because it's only applicable to the list function on the resource and there is no context of this once we reach `API.request`.

I'm not sure on the advantage of using a comma separated value over a list here as we're just splitting it back out into a list as soon as it's needed. Also this would make it slightly more involved for users who want to build this list dynamically as they now need to worry about joining it correctly as well.

> Note I think expanded: ["balance_transaction"] is ruby?

I'm not sure I follow here. The syntax in my examples are adding the `:expand` key into the existing `opts` keyword list.

snewcomer replied:
Oh I was looking at the PR desc w/o the braces around `expanded`.  I see in your test you meant for it to be a proper keyword list! 👍

snewcomer commented on lib/stripe/api.ex:
```diff
@@ -273,4 +275,15 @@ defmodule Stripe.API do
+
+    "#{url}?#{expand_str}"
+  end
+
```
could remove space and/or put above first function if you wanted.

swelham replied:
Ah, these spaces are being added by the formatter. Is this something you would prefer I bypass in this case or are you working towards formatting the code base?

snewcomer replied:
na no worries :)

swelham commented:
Thanks @snewcomer.

I have updated with tests for the changes in `Request` but was unsure how to test the changes in `API` as it looks there currently isn't anything in place for testing the requests that are sent.

snewcomer commented on test/stripe/request_test.exs:
```diff
@@ -0,0 +1,22 @@
+  describe "object expansion" do
+    test "prefix_expansions/2 should apply the given prefix to the expansion values" do
+      opts = [expand: ["balance_transaction"]]
+      request = Request.prefix_expansions(%Request{opts: opts}, "data")
```
nice tests!

swelham commented:
I haven't implemented expansion for the card endpoint as I didn't see them there! I was focused on everything within `core_resources`.

swelham commented:
I have now updated with the two extra tests and added the doc block for `prefix_expansions`. I wasn't sure how in-depth you would like those docs, so do let me know if you would like some deeper detail.

Many thanks for the thorough review 😃

snewcomer commented on lib/stripe/request.ex:
```diff
@@ -144,12 +150,43 @@ defmodule Stripe.Request do
+  be prefixed with `data.`. This is required because the stripe api nests the 
+  returned objects within `data: {}`.
+
+  For all `create`, `update`, `cancel` and `retrieve` functions this is not required.
```
what about `delete`?  I see you added prefix expansion on `delete_discount`?

snewcomer commented on lib/stripe/core_resources/customer.ex:
```diff
@@ -155,6 +155,7 @@ defmodule Stripe.Customer do
           {:ok, t} | {:error, Stripe.Error.t()}
   def delete_discount(id, opts \\ []) do
     new_request(opts)
+    |> prefix_expansions("data")
```
✂️  I _think_ this could be removed since this just returns a deleted boolean. ✂️

swelham replied:
Well spotted! This was supposed to be in the list function above, now update.

swelham commented:
Yes, both of those points would definitely be good improvements.

The second point will cause a stripe error so should be caught in the development, but based on the way the rest of the lib validates input data it would be a really nice to have on object expansion as well.

swelham commented:
Thanks for the feedback @begedin, I will try to find some time today to add that additional documentation.

snewcomer commented on README.md:
````diff
@@ -210,6 +210,43 @@ Create a connect standalone account. Grab your development `client_id`. Put it i
+However if we now include an expansion on the `balance_transaction` field using
+
+```elixir
+Charge.retrieve("ch_123", expand: ["balance_transaction"])
````
r /`expand: ["balance_transaction"]` / `[expand: ["balance_transaction"]]`

swelham replied:
The extra brackets here aren't required and are being removed by the formatter.

Is there an intention to use separate styling for this repo? If not I would be happy to submit a follow-up PR that applies formatting to the codebase if you would like?

snewcomer replied:
?  See your test here - https://github.com/code-corps/stripity_stripe/pull/393/files#diff-81a69829bd06582e30fdf8543ef74eb2R8

I think you need brackets around the 2nd argument -`[expand: ["balance_transaction"]]`

swelham replied:
In the test I am creating the keyword list separately to the function call and in this case the brackets are required. In the README example it's created inline and therefore optional (and automatically removed).

Here's the relevant docs from the Elixir style guide.

https://github.com/christopheradams/elixir_style_guide/blob/master/README.md#keyword-list-brackets

snewcomer replied:
woah.  Learn something new everyday!  🙏
